File: app/src/main/java/ui/activity/ShowNotificationActivity.java
package ui.activity;

import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.widget.NestedScrollView;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.button.MaterialButton;

import hlv.cute.todo.R;
import hlv.cute.todo.databinding.ActivityShowNotificationBinding;
import model.Notification;
import utils.ToastHelper;
import viewmodel.ShowNotificationViewModel;

public class ShowNotificationActivity extends BaseActivity {

    private ActivityShowNotificationBinding binding;

    private ShowNotificationViewModel viewModel;

    private ConstraintLayout toolbar;
    private NestedScrollView nested;
    private MaterialButton btnDone;

    private TextView txtTitle;
    private TextView txtCategory;
    private TextView txtLowPriority, txtNormalPriority, txtHighPriority;
    private ConstraintLayout lytDate, lytCategory;
    private TextView txtDateReminder, txtClockReminder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityShowNotificationBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        initViewModel();
        handleObserver();
        initViews();
        handleLogic();
    }

    private void initViewModel() {
        viewModel = new ViewModelProvider(this).get(ShowNotificationViewModel.class);
        viewModel.setIntent(getIntent());
    }

    private void handleObserver() {
        viewModel.getCloseLive().observe(this, mustClose -> {
            if (mustClose != null && mustClose)
                close();
        });
    }

    private void initViews() {
        binding.aImgBack.setOnClickListener(v -> close());
        binding.mBtnClose.setOnClickListener(v -> close());

        toolbar = binding.toolbar;
        nested = binding.nested;

        txtTitle = binding.txtTodoTitle;
        txtCategory = binding.txtCategory;

        txtLowPriority = binding.txtLowPriority;
        txtNormalPriority = binding.txtNormalPriority;
        txtHighPriority = binding.txtHighPriority;

        lytDate = binding.lytDate;
        lytCategory = binding.lytCategory;

        txtDateReminder = binding.txtDate;
        txtClockReminder = binding.txtClock;

        btnDone = binding.mBtnDone;
        handleShadowScroll();
    }

    private void handleShadowScroll() {
        nested.setOnScrollChangeListener(new NestedScrollView.OnScrollChangeListener() {
            final float dpShadow = getResources().getDimension(R.dimen.toolbar_shadow);

            @Override
            public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
                if (scrollY == 0) {
                    toolbar.animate().translationZ(0).setStartDelay(0).setDuration(200).start();
                } else if (scrollY > 50) {
                    toolbar.setTranslationZ(dpShadow);
                    toolbar.animate().translationZ(dpShadow).setStartDelay(0).setDuration(90).start();
                }
            }
        });
    }

    private void handleLogic() {
        Notification notif = viewModel.getNotification();
        if (notif == null) { //when manual close and open with home
            runActivity(MainActivity.class, true);
            return;
        }

        //todo: set shown true in startup get all is shown and delete it

        txtTitle.setText(notif.getContent());

        switch (notif.getPriority()) {
            case LOW:
            default:
                txtLowPriority.setVisibility(View.VISIBLE);
                txtNormalPriority.setVisibility(View.GONE);
                txtHighPriority.setVisibility(View.GONE);
                break;

            case NORMAL:
                txtLowPriority.setVisibility(View.GONE);
                txtNormalPriority.setVisibility(View.VISIBLE);
                txtHighPriority.setVisibility(View.GONE);
                break;

            case HIGH:
                txtLowPriority.setVisibility(View.GONE);
                txtNormalPriority.setVisibility(View.GONE);
                txtHighPriority.setVisibility(View.VISIBLE);
                break;
        }

        lytDate.setVisibility(viewModel.getLytDateVisibility());
        lytCategory.setVisibility(viewModel.getLytCategoryVisibility());

        if (viewModel.hasCategory())
            txtCategory.setText(notif.getCategory());

        if (viewModel.hasArriveDate()) {
            txtDateReminder.setText(viewModel.getDateReminder());
            txtClockReminder.setText(viewModel.getClockReminder());
        }


        btnDone.setOnClickListener(v -> {
            viewModel.done();
            ToastHelper.get().toast(getString(R.string.todo_done_successfully));
            new Handler().postDelayed(this::close, 500);
        });
    }

    private void close() {
        viewModel.deleteNotification();
        finish();
    }

    @Override
    public void onBackPressed() {
        close();
        super.onBackPressed();
    }

    @Override
    protected void onDestroy() {
        close();
        super.onDestroy();
    }
}

File: app/src/main/java/ui/activity/ShowNotificationActivity.kt
package ui.activity

import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import androidx.activity.viewModels
import androidx.core.widget.NestedScrollView
import com.aghajari.rlottie.AXrLottie
import com.aghajari.rlottie.AXrLottieDrawable
import dagger.hilt.android.AndroidEntryPoint
import hlv.cute.todo.R
import hlv.cute.todo.databinding.ActivityShowNotificationBinding
import model.Notification
import model.Priority
import ui.component.UiToolkit
import ui.component.bindingComponent.BaseViewBindingActivity
import utils.ToastUtil
import viewmodel.ShowNotificationViewModel
import javax.inject.Inject


@AndroidEntryPoint
class ShowNotificationActivity : BaseViewBindingActivity<ActivityShowNotificationBinding>() {

    override val bindingInflater: (LayoutInflater) -> ActivityShowNotificationBinding
        get() = ActivityShowNotificationBinding::inflate

    private val viewModel by viewModels<ShowNotificationViewModel>()

    @Inject
    lateinit var toastUtil: ToastUtil

    @Inject
    lateinit var uiToolkit: UiToolkit

    override fun initiate() {
        initViewModel()
        initViews()
        handleObserver()
    }

    private fun initViewModel() {
        viewModel.setIntent(intent)
    }

    private fun initViews() {
        AXrLottie.init(this)

        binding.aImgBack.setOnClickListener { close() }

        binding.mBtnClose.setOnClickListener { close() }

        binding.mBtnDone.setOnClickListener {
            binding.mBtnDone.isEnabled = false

            viewModel.done()

            toastUtil.successToast(provideResource.getString(R.string.todo_done_successfully_simple))

            binding.confetti.visibility = View.VISIBLE

            val width = uiToolkit.displayWidth
            val height = uiToolkit.displayHeight

            binding.confetti.lottieDrawable =
                AXrLottieDrawable
                    .fromAssets(this, "congratulations_confetti_lottie.json") //from assets
                    .setSize(width, height)
                    .setOnLottieLoaderListener(object : AXrLottieDrawable.OnLottieLoaderListener {
                        override fun onLoaded(drawable: AXrLottieDrawable?) {
                            Handler(Looper.myLooper()!!).postDelayed({ close() }, 3_000L)
                        }

                        override fun onError(drawable: AXrLottieDrawable?, error: Throwable?) {
                        }
                    })
                    .build()

            binding.confetti.playAnimation()
        }

        handleShadowScroll()
    }

    private fun handleShadowScroll() {
        binding.nested.setOnScrollChangeListener(object : NestedScrollView.OnScrollChangeListener {
            val dpShadow = provideResource.getDimen(R.dimen.toolbar_shadow)
            override fun onScrollChange(
                v: NestedScrollView,
                scrollX: Int,
                scrollY: Int,
                oldScrollX: Int,
                oldScrollY: Int
            ) {
                if (scrollY == 0) {
                    binding.toolbar.animate()
                        .translationZ(0f)
                        .setStartDelay(0)
                        .setDuration(200)
                        .start()

                } else if (scrollY > 50) {
                    binding.toolbar.translationZ = dpShadow
                    binding.toolbar.animate()
                        .translationZ(dpShadow)
                        .setStartDelay(0)
                        .setDuration(90).start()
                }
            }
        })
    }

    private fun handleObserver() {
        viewModel.closeLive.observe(this) { mustClose: Boolean ->
            if (mustClose) close()
        }

        viewModel.runMainLive.observe(this) { runMain: Boolean ->
            if (runMain)
                runActivity(MainActivity::class.java, true)
        }

        viewModel.notificationLiveData.observe(this) { notif: Notification? ->
            if (notif == null) { //when manual close and open with home
                runActivity(MainActivity::class.java, true)
                return@observe
            }

            //set shown true in startup get all is shown and delete it
            viewModel.setShown()
            binding.txtTodoTitle.text = notif.content

            when (notif.priority) {
                Priority.LOW -> {
                    binding.txtLowPriority.visibility = View.VISIBLE
                    binding.txtNormalPriority.visibility = View.GONE
                    binding.txtHighPriority.visibility = View.GONE
                }

                Priority.NORMAL -> {
                    binding.txtLowPriority.visibility = View.GONE
                    binding.txtNormalPriority.visibility = View.VISIBLE
                    binding.txtHighPriority.visibility = View.GONE
                }

                Priority.HIGH -> {
                    binding.txtLowPriority.visibility = View.GONE
                    binding.txtNormalPriority.visibility = View.GONE
                    binding.txtHighPriority.visibility = View.VISIBLE
                }

                else -> {
                    binding.txtLowPriority.visibility = View.VISIBLE
                    binding.txtNormalPriority.visibility = View.GONE
                    binding.txtHighPriority.visibility = View.GONE
                }
            }

            binding.lytDate.visibility = viewModel.getLytDateVisibility()
            binding.lytCategory.visibility = viewModel.getLytCategoryVisibility()

            if (viewModel.hasCategory())
                binding.txtCategory.text = notif.category

            if (viewModel.hasArriveDate()) {
                binding.txtDate.text = viewModel.getDateReminder()
                binding.txtClock.text = viewModel.getClockReminder()
            }
        }
    }

    private fun close() {
        viewModel.deleteNotification()
        finish()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        close()
    }

    override fun onDestroy() {
        super.onDestroy()
        close()
    }
}